adds timeouts to ODIS signer

alecps opened the pull request.

Commit:
adds timeouts to signer

codyborn commented on packages/phone-number-privacy/signer/src/server.ts:
```diff
@@ -52,14 +53,28 @@ export function createServer() {
 }
 
-async function callAndMeasureLatency(
+async function callWithTimeout(
```
I think we can set this at the Express level in index.ts
```
server.listen(port, () => {
    logger.info(`Server is listening on port ${port}`)
  }).setTimeout(1000)
  ```
  https://nodejs.org/dist/latest-v6.x/docs/api/http.html#http_server_settimeout_msecs_callback

nambrot replied:
+1

Though it would be nice in either approach to add a prometheus metric to track how many requests time out

alecps replied:
Yea my concern is that if we do it at the express level the latency measurement will never wrap up and we'll be left without insight into how many timeouts we're seeing

alecps replied:
What we'll see with this implementation when signers timeout is signers reporting latency measurements that hit the ceiling set by the timeout parameter. We'll also see a prometheus counter tracking the number of timeouts.

codyborn replied:
I think we should add this at the server level too as a last resort and set a higher limit

alecps replied:
👍

Commit:
moves timeout function to base package, adds signer timeout metric

Commit:
udpates dependency graph

Commit:
fixes typo

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into alecps/signerTimeouts

nambrot commented:
One thing that is interesting is that while we would timeout the request, express will not actually cancel the processing of it

https://github.com/expressjs/timeout

nambrot commented on packages/phone-number-privacy/signer/src/server.ts:
```diff
@@ -52,14 +54,23 @@ export function createServer() {
+  await timeout(handler, [req, res], config.timeout, timeoutRes)
+    .catch((error: any) => {
+      if (error === timeoutRes) {
+        Counters.signatureRequestsWithoutSessionID.inc()
```
Is this the counter you want to be incrementing?

nambrot commented on packages/phone-number-privacy/signer/src/signing/get-partial-signature.ts:
```diff
@@ -44,7 +44,7 @@ export async function handleGetBlindedMessagePartialSig(
   if (!request.body.sessionID) {
     logger.debug({ request: request.body }, 'Request does not have sessionID')
-    Counters.signatureRequestsWithoutSessionID.inc()
+    Counters.timeouts.inc()
```
Is this change intentional?

alecps replied:
Ah yea I mixed these two up

nambrot commented on packages/sdk/base/src/async.ts:
```diff
@@ -139,3 +139,29 @@ export async function concurrentValuesMap<IN extends any, OUT extends any>(
 }
+
+export const timeout = <T extends any[], U>(
+  inFunction: InFunction<T, U>,
```
Shouldn't this be a Promise?

alecps replied:
It could be..I guess it felt cleaner to create the promise inside the timeout function. It's also what the other functions here are doing and the code pattern was easier to copy. Let me know if you feel strongly about this

nambrot commented on packages/sdk/base/src/async.ts:
```diff
@@ -139,3 +139,29 @@ export async function concurrentValuesMap<IN extends any, OUT extends any>(
+  let timer: number
+  return Promise.race([
+    inFunction(...params),
+    new Promise(() => {
```
How does this promise ever resolve?

alecps replied:
setTimeout **rejects** (throws an error) to whatever you pass in as its 3rd argument when the timeout is reached. The Promise that wraps it also rejects to the same value, as does the Promise.race.

nambrot commented on packages/phone-number-privacy/signer/package.json:
```diff
@@ -18,6 +18,7 @@
     "ssl:keygen": "./scripts/create-ssl-cert.sh"
   },
   "dependencies": {
+    "@celo/base": "1.1.1-dev",
```
Will this work with our dockerfile building since its not available?

alecps replied:
hmm good question. I think this will also come up in Cody's PR

alecps replied:
Yea I think you're right. One workaround would be to import it in through the common package (which we're already doing in Cody's combiner timeout PR). We'd need to repubiish the package though

alecps replied:
Just pushed this change

nambrot commented on packages/sdk/base/src/async.ts:
```diff
@@ -139,3 +139,29 @@ export async function concurrentValuesMap<IN extends any, OUT extends any>(
   }, {})
 }
+
+export const timeout = <T extends any[], U>(
```
Maybe add some comment here on what the timeout behavior is. Right now in here it's not obvious that it would throw an error, but that's what the test seems to indicate

alecps replied:
will do

Commit:
fixes error

Commit:
adds backup timeout at express level

Commit:
addresses feedback

Commit:
import timeout through common package

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into alecps/signerTimeouts

Commit:
moves backup timeout to better spot

Commit:
removes commented out code

Commit:
parallelizes db ops

Commit:
adds TODOS

Commit:
adds timeout helper to sdk/base

Commit:
adds retryAsyncWithTimeout helper to sdk/base

Commit:
refactors retryAsyncWithBackOffAndTimeout() to use timeout()

Commit:
bumps package version and updates changelog

Commit:
adds docs

Commit:
Revert "bumps package version and updates changelog"

This reverts commit ed9040ea2cda7c6a84280b9ae841ccbd06619c96.

Commit:
Merge branch 'alecps/odisBase' of github.com:celo-org/celo-monorepo into alecps/signerTimeouts

Commit:
Adds timeout functions to sdk/base package (#7617)

### Description

In our efforts to improve ODIS scalability, the CAP team came across the need for these timeout async functions and they seemed like something that should be added to the base package. 

We've added retryAsyncWithBackOffWithTimeout() and timeout(). The former retries an async function up to a given number of times with a backoff between each retry and caps the attempt to call the function with a timeout. The timeout() function simply calls an async function and wraps the promise in a timeout.

Authorship credit goes to @codyborn who wrote the timeout logic in a separate PR 

### Tested

Unit tests added to async.test.ts

### Related issues

None

### Backwards compatibility

Yes

### Documentation

None

Commit:
add full node timeouts to signer

Commit:
adds db timeouts to signer

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into alecps/signerTimeouts

Commit:
adds dependency graph

codyborn committed:
ODIS Combiner Reliability (#7594)

### Description

- Adding timeouts to full-node requests
- “Failing open” when timeouts are hit
- Forcing outstanding request termination when k of m signatures have been collected
- Forcing outstanding request termination and returning early when k of m signatures are not possible

### Tested

- Test cases for new async method wrapper
- TODO: Deploy to alfajores and perform load tests to measure impact

### Backwards compatibility

Yes

### Documentation

N/A

Commit:
ODIS load test (#7602)

Integrates load test logic into monitor

Commit:
Merge branch 'odis/scalability' into alecps/signerTimeouts

Commit:
update dependency graph

Commit:
adds sdk/base to signer dockerfile (temporary)

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into alecps/signerTimeouts
